In [1]:
#User-based Recommendation Example using Cosine Similarity

In [2]:
import pandas as pd
import numpy as np

In [3]:
import os
os.getcwd()
os.chdir(r'C:\Users\officeuser\Downloads')

#### When we have the file which has the information in both text and numbers, we will need to use Encoding format - Latin 

##### As we find the cosine distance , its importatnt to have the conversion into a encoding format

In [4]:
#in this u.users data file the data is seprated by the | 

users = pd.read_csv('ml-100k/u.user', sep='|', names=['user_id','age','gender','occupation','zip_code'])

#in this data file the u.data file , data is seprated by the \t , if we dontuse sep it will all be one column

ratings = pd.read_csv('ml-100k/u.data', sep='\t', names=['user_id','item_id','rating','timestamp'])

#Encoding 

movies = pd.read_csv('ml-100k/u.item', sep='|', names=['movie_id','movie_title','release_date','video_release_date','IMDb URL','unknown','Action','Adventure','Animation','Children','Comedy','Crime','Documentary','Drama','Fantasy','Film-Noir','Horror','Musical','Mystery','Romance','Sci-Fi','Thriller','War','Western'], encoding='latin-1')
movies = movies.iloc[:,[0,1]]

In [5]:
users.head()

user_id  age gender  occupation zip_code
0        1   24      M  technician    85711
1        2   53      F       other    94043
2        3   23      M      writer    32067
3        4   24      M  technician    43537
4        5   33      F       other    15213

In [6]:
ratings.head()

user_id  item_id  rating  timestamp
0      196      242       3  881250949
1      186      302       3  891717742
2       22      377       1  878887116
3      244       51       2  880606923
4      166      346       1  886397596

In [19]:
movies.sample(10)

movie_id                                     movie_title
299        300                            Air Force One (1997)
835        836                                Ninotchka (1939)
1437      1438                                  Panther (1995)
1545      1546                         Shadows (Cienie) (1988)
653        654                                Chinatown (1974)
87          88                     Sleepless in Seattle (1993)
988        989                         Cats Don't Dance (1997)
722        723                         Boys on the Side (1995)
83          84  Robert A. Heinlein's The Puppet Masters (1994)
1540      1541               Beans of Egypt, Maine, The (1994)

#### Item ID and Movie ID are same both same metric, we can perform a join operation 

In [8]:
#The Right_on will remain and the left_on will combined to left_on

merge1 = pd.merge(movies,ratings,right_on='item_id',left_on='movie_id')
finalDF = pd.merge(merge1,users,left_on='user_id',right_on='user_id')
finalDF.head()

movie_id            movie_title  user_id  item_id  rating  timestamp  age  \
0         1       Toy Story (1995)      308        1       4  887736532   60   
1         4      Get Shorty (1995)      308        4       5  887737890   60   
2         5         Copycat (1995)      308        5       4  887739608   60   
3         7  Twelve Monkeys (1995)      308        7       4  887738847   60   
4         8            Babe (1995)      308        8       5  887736696   60   

  gender occupation zip_code  
0      M    retired    95076  
1      M    retired    95076  
2      M    retired    95076  
3      M    retired    95076  
4      M    retired    95076

In [9]:
ratings

user_id  item_id  rating  timestamp
0          196      242       3  881250949
1          186      302       3  891717742
2           22      377       1  878887116
3          244       51       2  880606923
4          166      346       1  886397596
...        ...      ...     ...        ...
99995      880      476       3  880175444
99996      716      204       5  879795543
99997      276     1090       1  874795795
99998       13      225       2  882399156
99999       12      203       3  879959583

[100000 rows x 4 columns]

In [10]:
len(ratings['user_id'].value_counts())

943

In [11]:
len(ratings['item_id'].value_counts())

1682

#### Bascially we are combining the common terms with the help of Pivot table. As one user has rated min 20 movies out of 943 we have. 1681 are differnt types of cx who have rated 943 movies, as we have many common we use Pivot table


In [12]:
# Association Matrix - Pivot Table
# pivotTable

learningMatrix = ratings.pivot_table(index=['item_id'],columns=['user_id'],values='rating').reset_index(drop=True)
learningMatrix.fillna(0,inplace=True)
learningMatrix

user_id  1    2    3    4    5    6    7    8    9    10   ...  934  935  936  \
0        5.0  4.0  0.0  0.0  4.0  4.0  0.0  0.0  0.0  4.0  ...  2.0  3.0  4.0   
1        3.0  0.0  0.0  0.0  3.0  0.0  0.0  0.0  0.0  0.0  ...  4.0  0.0  0.0   
2        4.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  4.0   
3        3.0  0.0  0.0  0.0  0.0  0.0  5.0  0.0  0.0  4.0  ...  5.0  0.0  0.0   
4        3.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
...      ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
1677     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1678     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1679     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1680     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1681     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   

user_id  937  938  939  940  941  942  943  
0        0.0  4.0  0.0  0.0  5.0  0.0  0.0  
1        0.0  0.0  0.0  0.0  0.0  0.0  5.0  
2        0.0  0.0  0.0  0.0  0.0  0.0  0.0  
3        0.0  0.0  0.0  2.0  0.0  0.0  0.0  
4        0.0  0.0  0.0  0.0  0.0  0.0  0.0  
...      ...  ...  ...  ...  ...  ...  ...  
1677     0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1678     0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1679     0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1680     0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1681     0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[1682 rows x 943 columns]

In [13]:
#apply Cosine Similarity Formula on Association Matrix
from scipy.spatial.distance import cosine
from scipy.spatial.distance import euclidean
from scipy.spatial.distance import hamming
from scipy.spatial.distance import minkowski


from sklearn.metrics import pairwise_distances

#Pairwise distance only accept Matrix object


movie_similarity = 1 - pairwise_distances( learningMatrix, metric="cosine" )
np.fill_diagonal( movie_similarity, 0 ) 

ratings_matrix = pd.DataFrame( movie_similarity )
ratings_matrix.head()

#movie_similarity object your model

0         1         2         3         4         5         6     \
0  0.000000  0.402382  0.330245  0.454938  0.286714  0.116344  0.620979   
1  0.402382  0.000000  0.273069  0.502571  0.318836  0.083563  0.383403   
2  0.330245  0.273069  0.000000  0.324866  0.212957  0.106722  0.372921   
3  0.454938  0.502571  0.324866  0.000000  0.334239  0.090308  0.489283   
4  0.286714  0.318836  0.212957  0.334239  0.000000  0.037299  0.334769   

       7         8         9     ...      1672  1673      1674      1675  \
0  0.481114  0.496288  0.273935  ...  0.035387   0.0  0.000000  0.000000   
1  0.337002  0.255252  0.171082  ...  0.000000   0.0  0.000000  0.000000   
2  0.200794  0.273669  0.158104  ...  0.000000   0.0  0.000000  0.000000   
3  0.490236  0.419044  0.252561  ...  0.000000   0.0  0.094022  0.094022   
4  0.259161  0.272448  0.055453  ...  0.000000   0.0  0.000000  0.000000   

       1676  1677  1678  1679      1680      1681  
0  0.035387   0.0   0.0   0.0  0.047183  0.047183  
1  0.000000   0.0   0.0   0.0  0.078299  0.078299  
2  0.032292   0.0   0.0   0.0  0.000000  0.096875  
3  0.037609   0.0   0.0   0.0  0.056413  0.075218  
4  0.000000   0.0   0.0   0.0  0.000000  0.094211  

[5 rows x 1682 columns]

In [21]:
try:
    user_inp=input('Enter the reference movie title based on which recommendations are to be made: ')
    inp=movies[movies['movie_title']==user_inp].index.tolist()
    
    inp=inp[0]
    
    # MVOIE SIMILARITY : has the equating similar inputs based on the user inputs

    movies['similarity'] = ratings_matrix.iloc[inp]
    movies.head(5)
    print("Recommended movies based on your choice of ",user_inp ,": \n", movies.sort_values( ["similarity"], ascending = False )[1:10])
    
except:
    print(movies.head(5))

Enter the reference movie title based on which recommendations are to be made: Sleepless in Seattle (1993)
Recommended movies based on your choice of  Sleepless in Seattle (1993) : 
      movie_id                     movie_title  similarity
392       393           Mrs. Doubtfire (1993)    0.676511
201       202            Groundhog Day (1993)    0.670296
215       216  When Harry Met Sally... (1989)    0.660843
731       732                     Dave (1993)    0.648957
203       204       Back to the Future (1985)    0.646832
738       739             Pretty Woman (1990)    0.628144
450       451                   Grease (1978)    0.625068
401       402                    Ghost (1990)    0.624408
81         82            Jurassic Park (1993)    0.624353


## ITs not that great as we have to give the movie name along with the year